
# **기상 데이터 및 선형회귀분석을 사용한 태양광 발전량 예측 모델**
# **(Prediction Model of Solar Power Generation Using Meteorological Data and Linear Regression Analysis)**



----
----

##**<목차>**

### **1. 라이브러리 및 데이터 setup**

(1) 라이브러리 불러오기

(2) CSV 파일 데이터 불러오기
             

### **2. 데이터 특성분석**
(1) 기술통계

(2) 상관분석
                        

### **3. 모델 구축 및 평가**

(1) 선형회귀모델 구축

(2) 모델 성능평가

----
----


### **1. 라이브러리 및 데이터 setup**

---
---

(1) 라이브러리 불러오기


[학습 목표]
* import를 사용한 라이브러리 불러오기, pip을 사용한 라이브러리 설치 방법 학습


[파이썬 함수, 모듈, 패키지, 라이브러리]
- 함수(function) : 특정 작업을 수행하는 독립된 기능을 수행하는 단위
- 모듈(module) : 함수, 변수, 클래스 등을 모아놓은 .py 확장자를 가진 파일
- 패키지(package) : 특정 기능과 관련된 여러 모듈들을 모아놓은 폴더
- 라이브러리(library) : 여러 패키지와 모듈들을 모아놓은 것

[라이브러리 불러오기]

        import 라이브러리명
        # 해당 모듈을 사용하겠다는 의미
        # '라이브러리명.모듈명' 등으로 호출

        import 라이브러리명 as 별칭
        # 해당 라이브러리를 별칭으로 사용하겠다는 의미
        # '별칭.모듈명' 등으로 호출

        from 라이브러리명 import 모듈명
        # 해당 라이브러리 안의 해당 모듈을 사용하겠다는 의미
        # '모듈명' 등으로 호출


[데이터 분석 관련 주요 라이브러리]

* 판다스(pandas): 데이터 조작 및 분석을 위한 라이브러리

        # pandas 불러오기
        import pandas as pd

* 넘파이(numpy) : 행렬이나 대규모의 다차원 배열을 쉽게 처리하기 위한 라이브러리

        # numpy 불러오기
        import numpy as np

* 맷플롯립(matplotlib) : 데이터를 차트(chart)나 플롯(plot)으로 시각화하기 위한 라이브러리

        # matplotlib.pyplot 불러오기
        import matplotlib.pyplot as plt


[라이브러리 설치방법]

      # 방법
      !pip intall 라이브러리명

      # 예시
      !pip install pandas
      !pip install numpy
      !pip install matplotlib



---
---

[연습문제 1]

판다스(pandas) 라이브러리를 불러오시오. (단, pd라는 별칭으로 불러올 것)

In [1]:
# 풀이
import pandas as pd

[자가진단 1-1]

넘파이(numpy) 라이브러리를 불러오시오. (단, np라는 별칭으로 불러올 것)

In [2]:
# 코딩
import numpy as np


[자가진단 1-2]

맷플롯립(matplotlib) 라이브러리의 주요 모듈인 pyplot를 불러오시오. (단, pyplot는 plt라는 별칭으로 불러올 것)

In [3]:
# 코딩
import matplotlib.pyplot as plt


[연습문제 2]

pip을 사용해서 pandas 라이브러리를 설치하시오.  

In [4]:
# 풀이
!pip install pandas

[자가진단 2]

pip을 사용해서 numpy, matplotlib 라이브러리를 설치하시오.

In [5]:
# 코딩
!pip install numpy
!pip install matplotlib


[형성평가 1]

pandas, numpy, matplotlib.pyplot 라이브러리를 불러오시오. (단, 각각 pd, np, plt라는 별칭으로 불러올 것)

In [6]:
# 코딩
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt


---
---

(2) CSV 파일 데이터 불러오기


[학습 목표]
* pandas 라이브러리를 사용해 CSV 파일을 불러와 데이터프레임으로 저장하는 방법과, 데이터프레임의 특성 및 사용방법 학습

[CSV 파일 불러오기]

        # 방법
        import pandas as pd
        변수명 = pd.read_csv('파일경로')   # 불러온 CSV 파일은 데이터프레임으로 저장됨

        # 예시
        import pandas as pd
        df = pd.read_csv('training_set.csv')

[데이터프레임]

        # 생성방법
        import pandas as pd
        df = pd.DataFrame({'temp':[23.4, 21.2, 19.3],
                           'humi':[74.5, 79.2, 86.9],
                           'co2': [602, 700, 791]})

        # 구조분석   

                                      columns          
                           
                                 temp	humi	co2
                            0	23.4	74.5	602
                 index      1	21.2	79.2	700
                            2	19.3	86.9	791

        # 컬럼(변수)명 추출
        col = list(df.columns)    # col = ['temp', 'humi', 'co2']

        # 특정 컬럼(변수)의 데이터 추출
        temp = list(df['temp'])   # temp = [23.4, 21.2, 19.3]
        humi = list(df['humi'])   # humi = [74.5, 79.2, 86.9]
        co2 = list(df['co2'])     # co2 = [602, 700, 791]

---
---



[연습문제 1]

아래와 같은 데이터프레임을 생성하여 df_ex에 저장한 후, df_ex에 저장된 내용을 확인해보시오.

* 데이터프레임

             0     1
        0  정재원  92
        1  박상길  72  

In [7]:
# 풀이
import pandas as pd

df_ex = pd.DataFrame([['정재원', 92], ['박상길', 72]]) # 변수명 = pd.DataFrame([데이터])
df_ex
#print(df_ex)

## 데이터 프레임 생성 시 구조 참고
# 변수명 = pd.DataFrame([리스트1, 리스트2, ... ,리스트3])

,0,1
0,정재원,92
1,박상길,72


[자가진단 1]

아래와 같은 데이터프레임을 생성하여 df에 저장한 후, df에 저장된 내용을 확인해보시오.

* 데이터프레임

            0     1     2
        0  A반  93.2  82.3
        1  B반  83.2  70.3
        2  C반  95.2  94.3

In [16]:
# 코딩
import pandas as pd

df = pd.DataFrame([['A반', 93.2, 82.3], ['B반', 83.2, 70.3], ['C반', 95.2, 94.3]])
df

,0,1,2
0,A반,93.2,82.3
1,B반,83.2,70.3
2,C반,95.2,94.3


[연습문제 2]

아래와 같은 데이터프레임을 생성하여 df_ex에 저장한 후, df_ex에 저장된 내용을 확인해보시오. (컬럼명이 변경되었음을 확인할 것)

* 데이터프레임

            이름  수학
        0  정재원  92
        1  박상길  72  

In [10]:
# 풀이 (1)
df_ex = pd.DataFrame([['정재원', 92], ['박상길', 72]], columns = ['이름','수학'])
df_ex

## 데이터 프레임 생성 시 구조 참고
# 변수명 = pd.DataFrame([리스트1, 리스트2, ... ,리스트3]], columns = [컬럼명1, 컬럼명2, ... 컬럼명3])

,이름,수학
0,정재원,92
1,박상길,72


In [ ]:
# 풀이 (2)
df_ex = pd.DataFrame({'이름':['정재원','박상길'], '수학':[92,72]})
df_ex

## 데이터 프레임 생성 시 구조 참고
# 변수명 = pd.DataFrame([{컬럼명1 : 리스트1, 컬럼명2 : 리스트2, ... 컬럼명n : 리스트n})

[자가진단 2]

아래와 같은 데이터프레임을 생성하여 df에 저장한 후, df에 저장된 내용을 확인해보시오. (컬럼명이 변경되었음을 확인할 것)

* 데이터프레임

           반   수학  국어
        0  A반  93.2  82.3
        1  B반  83.2  70.3
        2  C반  95.2  94.3

In [20]:
df.columns = ['반', '수학', '국어']
df


,반,수학,국어
0,A반,93.2,82.3
1,B반,83.2,70.3
2,C반,95.2,94.3


[연습문제 3]

연습문제 2에서 생성한 df_ex의 컬럼명을 추출하여 아래와 같은 리스트 데이터를 생성하시오.

* 리스트

       ['이름', '수학']

In [21]:
# 풀이(1)
col_ex_index = df_ex.columns
col_ex = list(col_ex_index)
col_ex

['이름', '수학']

In [22]:
# 풀이(2)
col_ex2 = list(df_ex)
col_ex2

['이름', '수학']

[자가진단 3]

자가진단 3에서 생성한 데이터프레임 df의 컬럼명을 추출하여 아래와 같은 리스트 데이터를 생성하시오.

* 리스트

       ['반', '수학', '국어']

In [23]:
# 코딩
col_df_index = df.columns
col_df = list(col_df_index)
col_df



['반', '수학', '국어']

[연습문제 4]

연습문제 2에서 생성한 데이터프레임 df_ex에서 '이름' 컬럼에 대한 데이터를 추출하여 아래와 같은 리스트 데이터를 생성하시오.

* 리스트

       ['정재원', '박상길']

In [24]:
# 풀이
x_series = df_ex['이름']
x = list(x_series)
x

# print(x_series)

['정재원', '박상길']

[자가진단 4]

자가진단 2에서 생성한 데이터프레임 df에서 '수학' 컬럼에 대한 데이터를 추출하여 아래와 같은 리스트 데이터를 생성하시오.

* 리스트

       [92.2, 83.2, 95.2]


In [26]:
# 코딩
math_scores = df['수학']
math_scores = list(math_scores)
math_scores


[93.2, 83.2, 95.2]

[연습문제 5]

test_set.csv 파일을 Colab에 업로드 한후, pandas를 사용하여 데이터프레임으로 df_tst에 저장한 후, df_tst에 어떤 데이터가 들어가 있는지 확인해보시오..

In [28]:
# 풀이
df_tst  = pd.read_csv('test_set.csv')
df_tst  = pd.read_csv('/content/test_set.csv')
df_tst

,power,temp_air,temp_gnd,humi,insol
0,5.7,14.1,14.3,76,0.36
1,3.5,12.0,13.1,98,0.18
2,0.0,11.0,12.2,100,0.01
3,1.7,10.9,11.6,100,0.00
4,0.8,10.7,11.1,100,0.00
5,5.8,9.9,9.8,95,0.08
6,12.2,10.3,11.6,97,0.39
7,24.8,11.6,15.5,93,1.08
8,32.4,12.4,16.7,87,1.54
9,5.0,14.0,17.1,80,1.65


[자가진단 5]

training_set.csv 파일을 Colab에 업로드 한후, pandas를 사용하여 데이터프레임 자료형으로 df에 저장한 후, df에 어떤 데이터가 들어가 있는지 확인해보시오.

In [31]:
# 코딩
df = pd.read_csv('/content/training_set.csv')
df


,power,temp_air,temp_gnd,humi,insol
0,5.4,4.6,2.6,100,0.10
1,17.4,8.8,6.1,73,0.69
2,37.9,10.5,12.3,65,1.41
3,54.7,12.2,19.8,61,2.06
4,76.5,12.7,26.2,59,2.68
...,...,...,...,...,...
114,18.4,14.9,14.7,73,0.56
115,15.9,14.7,15.3,74,0.64
116,20.4,15.0,17.2,72,0.79
117,16.5,15.2,17.0,67,0.73


[형성평가 1]

pandas 라이브러리를 pd라는 별칭으로 불러오시오.

In [32]:
# 코딩
import pandas as pd



[형성평가 2]

training_set.csv 파일에 저장된 데이터를 불러와 데이터프레임으로 df에 저장하시오.

In [34]:
# 코딩
df = pd.read_csv('/content/training_set.csv')


[형성평가 3]

데이터프레임 df에서 컬럼명을 추출하여 아래와 같은 리스트 데이터를 생성하시오.

* 리스트
       col = ['power', 'temp_air', 'temp_gnd', 'humi ', 'insol']


In [37]:
# 코딩
df_index = list(df.columns)
df_index

['power', 'temp_air', 'temp_gnd', 'humi', 'insol']

[형성평가 4]

데이터프레임 df의 각 컬럼에 대한 데이터를 추출하여 아래와 같은 리스트 데이터를 생성하시오.

* 리스트

       power = [5.4, 17.4, 37.9, 54.7, 76.5, 83.5, 68.0, 79.6, 69.8, 53.6, ...]
       temp_air = [4.6, 8.8, 10.5, 12.2, 12.7, 14.1, 14.4, 15.3, 14.8, 14.5, ...]
       temp_gnd = [2.6, 6.1, 12.3, 19.8, 26.2, 29.6, 29.8, 31.4, 27.8, 25.0, ...]
       humi = [100, 73, 65, 61, 59, 54, 51, 45, 46, 58, ...]
       insol = [0.1, 0.69, 1.41, 2.06, 2.68, 2.52, 2.27, 2.65, 2.16, 1.55, ...]

In [45]:
# 코딩
power = list(df['power'])
temp_air = list(df['temp_air'])
temp_gnd = list(df['temp_gnd'])
humi = list(df['humi'])
insol = list(df['insol'])

print('power =', power)
print('temp_air =', temp_air)
print('temp_gnd =', temp_gnd)
print('humi =', humi)
print('insol =', insol)


power = [5.4, 17.4, 37.9, 54.7, 76.5, 83.5, 68.0, 79.6, 69.8, 53.6, 32.1, 5.6, 4.8, 19.0, 32.1, 48.3, 62.0, 66.6, 76.2, 74.5, 47.9, 37.6, 27.8, 10.8, 6.0, 17.6, 34.8, 53.6, 33.6, 81.6, 90.7, 29.3, 42.4, 38.1, 31.0, 7.0, 5.8, 21.3, 42.7, 58.7, 73.0, 79.3, 83.2, 73.0, 75.8, 60.1, 39.4, 7.6, 5.0, 18.2, 37.6, 55.4, 66.8, 75.7, 73.7, 83.6, 74.4, 58.9, 39.4, 8.2, 0.8, 6.4, 18.1, 32.9, 47.5, 51.1, 64.7, 73.4, 61.2, 54.4, 47.7, 23.7, 8.3, 5.4, 16.3, 35.1, 48.7, 65.7, 62.5, 65.1, 62.5, 64.6, 51.7, 24.5, 2.1, 0.7, 1.3, 1.3, 4.6, 16.7, 3.9, 6.1, 12.1, 86.1, 71.3, 17.8, 28.8, 10.8, 4.3, 21.3, 40.4, 63.6, 77.4, 85.5, 87.0, 87.6, 74.9, 61.0, 40.3, 12.5, 0.9, 5.7, 18.2, 13.3, 18.4, 15.9, 20.4, 16.5, 11.6]
temp_air = [4.6, 8.8, 10.5, 12.2, 12.7, 14.1, 14.4, 15.3, 14.8, 14.5, 13.5, 12.2, 4.8, 6.4, 7.8, 9.7, 11.6, 13.1, 13.9, 14.6, 13.8, 13.2, 12.8, 10.9, 2.6, 6.3, 10.0, 12.8, 13.3, 14.6, 14.9, 14.3, 14.9, 14.3, 14.1, 13.0, 6.4, 8.1, 9.8, 11.0, 12.4, 13.7, 15.8, 17.5, 18.3, 18.8, 15.7, 14.0, 6.4, 8.7, 1